Assign CPUset for system reserved cgroup.

ipuustin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test

/kind feature
> /kind flake

**What this PR does / why we need it**:

Previously system reserved cgroup could not be declared if `--reserved-cpus` option was used. Now it's allowed, and the cpuset defined with `--reserved-cpus` will be assigned to the system reserved cgroup if `--enforce-node-allocatable=system-reserved`.

So, in order to let system services (assigned to cpuset /foobar.slice on a systemd system) on CPUs 4 and 5, you could run kubelet with

    kubelet --reserved-cpus=4,5 \
            --enforce-node-allocatable=system-reserved \
            --system-reserved-cgroup=foobar

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #85764

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Kubelet can use --system-reserved-cgroup with --reserved-cpus option.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

Documentation PR will need to be submitted when the command line interface change is confirmed.

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

ipuustin committed:
Assign CPUset for system reserved cgroup.

Previously system reserved cgroup could not be declared if
--reserved-cpus option was used. Now it's allowed, and the
cpuset defined with --reserved-cpus will be assigned to the system
reserved cgroup if --enforce-node-allocatable=system-reserved.

So, in order to let system services (assigned to cpuset /foobar.slice on
a systemd system) on CPUs 4 and 5, you could run kubelet with

    kubelet --reserved-cpus=4,5 \
            --enforce-node-allocatable=system-reserved \
            --system-reserved-cgroup=foobar

mattjmcnaughton reviewed (commented):
@ipuustin thanks for your pr :)

I just want to be sure I'm following the larger context here - is this diff intended to implement https://github.com/kubernetes/enhancements/pull/1319/files in its entirety? Or is it tangential?
I think its the latter, but I just want to be sure I'm not making any incorrect assumptions.

ipuustin commented:
@mattjmcnaughton It's just tangential. I think the KEP doesn't address this use case. I just linked it because it's clear that the KEP is related and this PR is part of the "feature space" that the KEP discusses.

mattjmcnaughton commented:
> @mattjmcnaughton It's just tangential. I think the KEP doesn't address this use case. I just linked it because it's clear that the KEP is related and this PR is part of the "feature space" that the KEP discusses.

Gotcha gotcha - thanks!

klueska commented:
I agree with the overall idea. I need to spend more time looking at the code and thinking about the approach though.

/ok-to-test

ipuustin commented:
@klueska: This PR doesn't deal with `--kube-reserved=` or even `--system-reserved=` flags at all -- those are already handled by the container manager code, which adds cpu shares to the corresponding cgroups according to the cpu reservation values. The missing part was the cpuset which can now be set with `--reserved-cpus`.

This said, this PR might add some interaction between `--system-reserved=` and `--reserved-cpus=`, because I think that the system reserved value is derived from the cpuset value if `--reserved-cpus=` is used, so after this PR both the cpu shares corresponding to the cpuset and the cpuset itself are set on the system cgroup.

ipuustin commented:
/retest

jianzzha commented:
I would like to validate this idea. The option "--reserved-cpus="  works with isolcpu kernel option as interim solution, using cgroup would be the next step. How would you vision the overrall cgroup structure? or more specifically, how would kubelet command line looks like with the cgroups?

ipuustin commented:
@jianzzha The current docs say:

> reserved-cpus is meant to define an explicit CPU set for OS system daemons and kubernetes system daemons. This option is added in 1.17 release. reserved-cpus is for systems that do not intent to define separate top level cgroups for OS system daemons and kubernetes system daemons with regard to cpuset resource.

The idea in this PR is to let systems define the top-level cgroup for OS system daemons and automatically restrict the cpuset there according to reserved-cpus setting. In effect the last sentence above would be replaced something like this:

> You can optionally use reserved-cpus with a system cgroup. In this configuration the OS system daemons will be restricted to run only on cpus present in the reserved-cpus cpuset.

It is actually a good question whether or not the same semantics should follow for `kube-reserved` cgroup, too. We could let the `--kube-reserved-cgroup` be restricted to the reserved-cpus cpuset, if both are present on the command line, in exactly the same way as this PR proposes for the `system-reserved` cgroup. What do you think?

jianzzha commented:
potential challenge of having a cpuset cgroup for system-reserved and enforcing it:
all systemd service cgroup need to be under this parent cgroup.  Since some of them starts before kubelet and have already have their individual cpuset.cpus set;  when kubelet starts and enforce this on the parent setting, it needs to reset each of the individual service unit'  cpuset.cpus first. Have you tested your code and see if this actually happens?

jianzzha commented:
I did a quick test and yes my concern is valid - if there is already other system service unit under this system-reserved-cgroup, the simple update on this system-reserved-cgroup will not work.

To make it work, update all pre-existing systemd service  under the cpuset cgroup is necessary. However there might be danger of doing that - what if some service unit has its own cpuset setting for whatever purpose that it wants to keep.

For now We did have a workaround as described in  https://github.com/kubernetes/kubernetes/issues/85764

ipuustin commented:
@jianzzha If I understand correctly, you mean a scenario like this:
```
└─system.slice
  ├─foobar.service
  │ └─1019 /sbin/foobar
```
Here `foobar.service` would have been set by systemd to have (for example) a cpuset of `0-5`, and then kubelet would try to set the `system.slice` cpuset to be (for example) `4-6` and fail because the top-level cpuset needs to include the cpus in all the leaf cpusets.

I think the right solution here would be to error out with a good error message: the two configuration sources (systemd and kubelet) are conflicting. The problem is that in order for the system to function as needed, the requested system-reserved cpuset cannot be used, and the administrator needs to find out what's the reason for that and how it should be solved.

jianzzha commented:
> ```
> └─system.slice
>   ├─foobar.service
>   │ └─1019 /sbin/foobar
> ```
> 
> Here `foobar.service` would have been set by systemd to have (for example) a cpuset of `0-5`, and then kubelet would try to set the `system.slice` cpuset to be (for example) `4-6` and fail because the top-level cpuset needs to include the cpus in all the leaf cpusets.

Yes, this is what I meant.

> I think the right solution here would be to error out with a good error message: the two configuration sources (systemd and kubelet) are conflicting. The problem is that in order for the system to function as needed, the requested system-reserved cpuset cannot be used, and the administrator needs to find out what's the reason for that and how it should be solved.

I agree that whoever chooses to let  kubelet  to enforce system cgroup setting should understand the consequence and knows how to read into the log message if kubelet fails.

2rs2ts commented:
FYI, my company would probably want to be able to specify the system-reserved cpus separately from the kube-reserved cpus, so I think a final design should account for that

ipuustin committed:
Assign --reserved-cgroup cpuset also to kube-reserved cgroup.

ipuustin commented:
I added code for handling the `--kube-reserved-cgroup` the same way how the `--system-reserved-cgroup` is handled. @2rs2ts, you should be able to handle your use case by ` --enforce-node-allocatable=system-reserved,kube-reserved` flag: if you do not set `kube-reserved` there, then the cpuset is not set for it.

2rs2ts commented on cmd/kubelet/app/server.go:
```diff
@@ -677,15 +677,17 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, featureGate f
+				s.KubeReserved = make(map[string]string)
+			}
 			s.SystemReserved["cpu"] = strconv.Itoa(reservedSystemCPUs.Size())
+			s.KubeReserved["cpu"] = strconv.Itoa(reservedSystemCPUs.Size())
```
I think, ideally, you would have these cpusets be discrete, no?

ipuustin replied:
That would definitely provide more options to the system administrator -- the kubernetes components and the system components could be separated on the system level. However, at the moment we only have the `--reserved-cpus` flag to indicate cpuset not available to pods.

2rs2ts replied:
So why not have `--system-reserved-cpus` and `--kube-reserved-cpus`?

ipuustin commented on pkg/kubelet/cm/node_container_manager_linux.go:
```diff
@@ -130,15 +131,16 @@ func (cm *containerManagerImpl) enforceNodeAllocatableCgroups() error {
+		return fmt.Errorf("%q cgroup is not configured properly", cgroupConfig.Name)
 	}
-	klog.V(4).Infof("Enforcing limits on cgroup %q with %d cpu shares, %d bytes of memory, and %d processes", cName, cgroupConfig.ResourceParameters.CpuShares, cgroupConfig.ResourceParameters.Memory, cgroupConfig.ResourceParameters.PidsLimit)
+	cgroupConfig.ResourceParameters.CpuSet = cpus
```
That's right, if `--enforce-node-allocatable=system-reserved,kube-reserved` flag is set.

ipuustin commented:
/retest

ipuustin committed:
test: fix error logging.

In case the test case succeeded, the test called panic() instead of
reporting the actual error.

ipuustin committed:
test: change test case with reserved cpus and system cgroup to succeed.

ipuustin commented:
/retest

ipuustin commented:
/retest

klueska commented:
/assign @klueska

Hey @ipuustin, sorry for the *long* delay on getting back to you on this.

After reading through the comments and thinking about this for a while, I have the following comments / suggestions:

1) It feels a bit strange to arbitrarily decide that the `--reserved-cpus` flag should *only* be linked  to `system-reserved` CPUs  such that setting `--enforce-node-allocatable=system-reserved` will set the `cpuset` for these CPUs on the `system-reserved-cgroup`.

2) In the same vein, it  would be just as strange to arbitrarily decide that the `--reserved-cpus` flag should *only* be linked  to `kube-reserved` CPUs  such that setting `--enforce-node-allocatable=kube-reserved` will set the `cpuset` for these CPUs on the `kube-reserved-cgroup`.

3) It feels equally as strange to arbitrarily decide that combining them (as you have done in your most recent patch) to say that `--reserved-cpus`  links to both `system-reserved` *and* `kube-reserved` such that the same `cpuset` will be set up for both of them if specified as part of `--enforce-node-allocatable`.

As such, I feel like the only way to deal with this properly would be to introduce separate `--system-reserved-cpus` and `--kube-reserved-cpus` such that:

1) A user has *not* specified `--reserved-cpus` (which, as the documentation states, always takes precedence over any kube-reserved or kube-reserved settings). 
2) The number of CPUs listed in `--system-reserved-cpus` and `--kube-reserved-cpus` *matches* the CPU count specified in `system-reserved` and `kube-reserved` respectively.

I still think having introduced `--reserved-cpus` on its own is useful for cases where the sysadmin really wants to manage these CPUs himself (and not use system-reserved or kube-reserved at all). But trying to arbitrarily map `--reserved-cpus` onto either `system-reserved`, `kube-reserved` (or both )doesn't fell like a good approach.

klueska commented:
/cc @derekwaynecarr

for his thoughts

ipuustin commented:
Thanks @klueska for your comments!

I think the main point of this effort is automation of the cgroup cpuset settings. I believe that currently people are using `--reserved-cpus=` cpuset so that they assign the system services (and kube services) to that cpuset manually (to keep the guaranteed pods allocated to free cpus). The point of this PR is to automate the effort.

Adding new command line options (`--system-reserved-cpus` and `--kube-reserved-cpus`) definitely helps to achieve three goals:
 1. Being able to allocate a separate cpuset to kube services. I can think of a situation where the user might want to run those services on a separate cpuset than the system daemons, especially if there is some negative interaction between them.
 2. There is no need to touch the semantics of the existing command line options.
 3. Explicit is better than implicit. 

However, the problems I see are:
 1. Unclear semantics for `--enforce-node-allocatable=`. Is it needed for the new options to take effect or not? If not, why does it work differently?
 2. Confusion with the existing `--reserved-cpus=` option. In a way, this will mean that `--reserved-cpus=` option just "takes away" cpus for later manual assignment by the user, while `--system-reserved-cpus=` takes them away and assigns them right away to the system cgroup. A natural question is to ask what happens if `--reserved-cpus=` is set at the same time as `--system-reserved-cgroup=`, and why doesn't that work.
 2. Adding more kubelet options slowly leads to an combinatoric explosion -- we need to understand the semantics of all combinations (whether they are allowed and what do they mean) and possibly "recommend" some settings for various use cases.

I plan to go forward with this by implementing your suggestion to try it out in practice. It's easier to discuss the concepts with the code,

milieu commented:
I don't want this PR to go stale like the issue too >.<

milieu commented:
/reopen

fromanirh commented:
/reopen
/remove-lifecycle rotten

wgahnagl commented:
/triage accept

dims commented:
Is this PR still needed, please rebase if so (or we can close it?)

2rs2ts commented:
The underlying issue still definitely exists, but we need the author to come rebase and address PR feedback

ipuustin commented:
This PR is badly outdated and it would probably be the best to start from scratch with @klueska 's proposal above. I don't think I'll have a chance to work on this in the near future unfortunately. Also it's indeed clear that the effort at https://github.com/kubernetes/kubernetes/pull/103933 is related to this. I'm closing this one so that the focus can be on the new work in that PR.
